File: zad1.cpp
#include <iostream>
#include <vector>
#include <map>
#include <iomanip>
#include <windows.h>
#include <algorithm>

using namespace std;

struct item { // продукт
    string name;
    int quantity;
};

class sklad { // склад
private:
    map<string, item> storage; // хранилище
    vector<char> zonees; // список зон
    
public:
    void add(string address, string name, int quantity) { // ф-и€ добавлени€ товара
        if (storage.count(address) != 0) { // проверка есть ли така€ €чейка
            if (storage[address].name != name && storage[address].name != "") { // проверка, зан€ты ли €чейка другим продуктом
                cout << "ќшибка, €чейка зан€та!" << endl;
            } else {
                if (storage[address].quantity + quantity <= 10) { // проверка, превышает ли кол-во продуктов при добавлении число 10
                    storage[address] = {name, storage[address].quantity + quantity};
                } else {
                    cout << "ќшибка, не хватает места в €чейке!" << endl;
                }
            }
        } else { // если нет €чейки, то ошибка
            cout << "ќшибка, нет такой €чейки!" << endl;
        }
    }

    void remove(string address, int quantity) { // ф-и€ удалени€ товара
        if (storage.count(address) != 0) { // проверка есть ли така€ €чейка
            if (storage[address].quantity - quantity == 0) { // проверка, если при удалении продуктов стало 0, то освободим €чейку от продукта
                storage[address] = {"", 0};
            } else if (storage[address].quantity - quantity > 0) { // проверка, если при удалении продуктов их будет > 0, то выполн€ем
                storage[address].quantity -= quantity;
            } else { // если < 0, то продуктов не хватает
                cout << "ќшибка, не хватает продуктов дл€ удалени€!" << endl;
            }
        } else {
            cout << "ќшибка, нет такой €чейки!" << endl;
        }
    }

    void info () { // ф-и€ вызова информации
        cout << "јдресс\t" << "ѕродукт\t" << " ол-во" << endl; // информаци€ по загруженным €чейкам
        float countzanyato = 0.0;
        float countzonezanyato;
        for (auto i : storage) {
            if (i.second.quantity != 0) { // вывод непустых €чеек
                cout << i.first << "\t" << i.second.name << "\t" << i.second.quantity << endl;
                countzanyato++;
            }
        }

        float countvsego = 0.0; // информаци€ по общей загруженности склада
        cout << "«агруженность склада: ";
        for (auto i : storage) { // считаем всего €чеек
            countvsego++;
        }
        float zagrsklad = (countzanyato/countvsego)*100; // высчитываем загруженность склада
        cout << zagrsklad << "%" << endl;

        float countvsegozones = countvsego / zonees.size(); // »нформаци€ по общей загруженности каждой зоны
        vector<float> countzanyatozones(zonees.size());
        cout << "«агруженность каждой зоны:" << endl;
        for (int i = 0; i < zonees.size(); ++i) {
            for (auto j : storage) {
                if(j.first[0] == zonees[i] && j.second.quantity > 0) { // считаем зан€тые зоны
                    countzanyatozones[i]++;
                }
            }
        }
        float zagrzones;
        for (int i = 0; i < countzanyatozones.size(); ++i) {
            cout << "«она " << zonees[i] << ": ";
            zagrzones = (countzanyatozones[i]/countvsegozones) * 100;
            cout << zagrzones << "%" << endl;
        }

        cout << "ѕустые €чейки: "; // информаци€ по пустым €чейкам
        for (auto i : storage) {
            if (i.second.quantity == 0) {
                cout << i.first << " ";
            }
        }
        cout << endl;
    }

    void GenerationAddress() { // формируем адресса склада
        vector<char> zones = {'A', 'B', 'C'};
        zonees = zones;
        for (auto i : zones) {
            for (int polka = 1; polka <= 5; ++polka) { // проходим по каждой полке
                for (int section = 1; section <= 5; ++section) { // проходим по каждой секции
                    for (int stelaj = 1; stelaj <= 3; ++stelaj) { // проходим по каждому стелажу
                        // —оздаем адресса
                        stringstream s;
                        s << i << stelaj << section << polka;
                        storage[s.str()] = {"", 0};
                        
                    }
                }
            }
        }
    }
};

int main(){
    SetConsoleCP (1251);
    SetConsoleOutputCP (1251);

    sklad warehouse;
    item tovar;
    string address;

    warehouse.GenerationAddress(); // формируем склад(зоны A/B/C, 3 стелажа, 5 секций, 5 полок)

    cout << "¬ыберите команду: ADD, REMOVE, INFO, EXIT дл€ выхода: "; // работа с командами
    string command;
    bool k = true;
    while (k == true) {
        cin >> command;
        if (command == "ADD") {
            cin >> address >> tovar.name >> tovar.quantity;
            warehouse.add(address, tovar.name, tovar.quantity);
        } else if (command == "REMOVE") {
            cin >> address >> tovar.quantity;
            warehouse.remove(address, tovar.quantity);
        } else if (command == "EXIT") {
            cout << "¬ыход из программы..." << endl;
            k = false;
        } else if (command == "INFO") {
            warehouse.info();
        } else {
            cout << "ќшибка команды" << endl;
        }
    }

    return 0;
}

File: zad1.py
class Item:
    def __init__(self, name='', quantity=0):
        self.name = name
        self.quantity = quantity

class Sklad:
    def __init__(self):
        self.storage = {}
        self.zones = []

    def add(self, address, name, quantity):
        if address in self.storage:
            if self.storage[address].quantity + quantity <= 10:
                self.storage[address] = Item(name, self.storage[address].quantity + quantity)
            else:
                print("Ошибка, не хватает места для продуктов")
        else:
            print("Такого адреса не существует")

    def remove(self, address, quantity):
        if address in self.storage:
            if self.storage[address].quantity - quantity >= 0:
                self.storage[address].quantity -= quantity
            else:
                print("Нет стольки продуктов для удаления")
        else:
            print("Такого адреса не существует")

    def info(self):
        print("Адресс\tПродукт\tКол-во")
        count_occupied = 0.0
        for address, item in self.storage.items():
            if item.quantity != 0:
                print(f"{address}\t{item.name}\t{item.quantity}")
                count_occupied += 1

        total_count = len(self.storage)
        print("Загруженность склада: {:.2f}%".format((count_occupied / total_count) * 100))

        total_zones = total_count / len(self.zones)
        occupied_zones = [0.0] * len(self.zones)
        print("Загруженность каждой зоны:")
        for i, zone in enumerate(self.zones):
            for address, item in self.storage.items():
                if address[0] == zone and item.quantity > 0:
                    occupied_zones[i] += 1
            print("Зона {}: {:.2f}%".format(zone, (occupied_zones[i] / total_zones) * 100))

        print("Пустые ячейки: ", end='')
        for address, item in self.storage.items():
            if item.quantity == 0:
                print(address, end=' ')
        print()

    def generate_addresses(self):
        zones = ['A', 'B', 'C']
        self.zones = zones
        for zone in zones:
            for shelf in range(1, 6):
                for section in range(1, 6):
                    for rack in range(1, 4):
                        address = f"{zone}{rack}{section}{shelf}"
                        self.storage[address] = Item()

warehouse = Sklad()
warehouse.generate_addresses()

print("Выберите команду: ADD, REMOVE, INFO, EXIT для выхода:")
k = True
while k:
    command = input()
    if command == "ADD":
        address, name, quantity = input().split()
        quantity = int(quantity)
        warehouse.add(address, name, quantity)
    elif command == "REMOVE":
        address, quantity = input().split()
        quantity = int(quantity)
        warehouse.remove(address, quantity)
    elif command == "EXIT":
        print("Выход из программы...")
        k = False
    elif command == "INFO":
        warehouse.info()
    else:
        print("Ошибка команды")